text/2022-07-04-subsetting-experiments.md

```diff
@@ -0,0 +1,169 @@
+  general-purpose font compiler
+- the code is (hopefully) easier to follow; all of the subsetting code
+  itself just involves modifying various types in-place ([see example][subset-impl])
+- as each table owns its data, and subsetting mutates the table in place
+  (not writing to a shared buffer) multiple tables can be subset in
+      parallel.
+
+### note on a design more similar to `hb-subset`
+
+The design I've chosen here makes clear trade-offs; in particular it is trading
+off some speed in favour of improved ergonomics and robustness (it relies on
+much less hand-written code).
+
+If this trade-off ends up being unacceptable, it will always be possible to
+write a more bespoke implementation, which would operate directly on the parse
+types, and write out bytes directly, very similar to how `hb-subset` works. This
+would be more work and more error-prone, but it is a viable escape hatch if
+needed. In a final implementation, it would even be possible to do this
+*selectively*, for instance by doing these manual implementations only for
+GPOS/GSUB, and still using the compile types for less complicated tables.
+
+## benchmark setup
+
+The goal for this test was not to build a production subsetter; rather it was to
+build the 'simplest credible comparison', that is to implement exactly as much
+functionality as is necessary in order to get a ballpark sense of performance.
+
+After chatting with @garret, we decided on a simple comparison: subsetting using
+glyph ids, dropping all tables but GPOS, and testing with a font that doesn't
+use complex (contextual) shaping rules. This saves us the work of implementing
+subsetting for the complex tables, and also saves the work of having to compute
+the glyph closure, since we drop all the tables that might be involved.
+
+The invocation to `hb-subset` looks like,
+
+```sh
+> ./hb-subset LibreBodoni-Regular.ttf -o subset_hb.ttf --gids 5,10,22,50-400 \
+  --drop-tables="*" --drop-tables-=GPOS -n 20
+```
+
+On the rust side, we have a simple binary that tries to emulate this behaviour:
+it drops all tables but GPOS, which it subsets based on some set of glyph ids.
+
+The rust binary is [avaialable here][subset_gpos], and the invocation looks
+like:
+
+```sh
+> ./subset_gpos LibreBodoni-Regular.ttf -o subset_rs.ttf --gids 5,10,22,50-400 --runs 20
+```
+
+## results
+
+> Initial disclaimer: running each of these commands produces *similar* binary output:
+there are an equal number of subtables with an equal number of records,
+equal-length coverage tables, etc. The output is not identical; for instance
+there seems to be a difference in the mapping of old->new glyph ids (presumably
+the result of off-by-ones in my impl somewhere) but I do not believe that these
+differences are important for evaluating performance.
+
+Timing the execution of the commands above, running each five times:
+
+```
+hb:
+0.01s user 0.01s system 66% cpu 0.035 total
```

Comment:
I would suggest significantly increasing the number of iterations until the run time is at least in the 5-10 seconds range so you can reduce the impact of setup overhead (such as loading the binary and font from disk).

Another option is to use the subsetting benchmark suite here: https://github.com/harfbuzz/harfbuzz/blob/main/perf/benchmark-subset.cc

Reply:
Okay, good to note.

I'm slightly of two minds here: on the one hand increasing the number of iterations definitely makes the benchmark more representative of the "interesting" work; but on the other hand, if setup overhead is some significant fraction of runtime, and we're ultimately trying to compare theoretical end-to-end time (i.e. receiving an http request and returning a response) then it doesn't seem unreasonable to want to include that?


(but maybe I'm just being grumpy, because `n` up to 10k or so makes harfbuzz 2-3x faster, which dulls my previous excitement somewhat)

Reply:
My experience with the subsetter is that the slowest parts are algorithmic (the glyph closure of Noto Nastaliq Urdu is currently the slowest), so I wouldn't worry much about what you are seeing right now; the serialization / deserialization part is fast enough that a 2/3x factor in C++/rust performance is not going to be an issue IMO.

Reply:
The overhead of running it on the command line is not going to be representative of overhead typical usages like in a web server. In those cases the font to be subset will likely already be in memory and then the subsetting time will be dominated primarily by what's within the actual subsetting call (which is what you'll be measuring with a high number of iterations).

Agree with Behdad, this is currently looking at a very specific slice of the overall work which would happen in a full subsetting operation. Other parts will likely have different relative performances. So it's probably difficult to extrapolate to what a full subset would look like just yet.

Also there's definitely a "fast enough" threshold. Being 2-3x slower isn't a big deal if you're going from 0.5 ms to 1.5ms in subsetting time. 2-3x slowdown is more significant for subsets that are in the 10ms range, but as Behdad noted those times are mostly about closure and repacking which could end up being competitive to the hb implementations since they rely far less on the serialization mechanism being tested here.